### PR TITLE
project_loader: add build-environment part property

### PR DESCRIPTION
**Pull request by kyrofa**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

Beyond using a custom plugin, there is no way for a part to specify environment variables to be defined during the part's lifecycle. This PR resolves [LP: #1778174](https://bugs.launchpad.net/snapcraft/+bug/1778174) by adding a new property to support doing so declaratively in the snapcraft.yaml, called `build-environment`. Its usage looks like this:

    parts:
      my-part:
        plugin: <plugin name>
        build-environment:
          - VARIABLE1_NAME: VARIABLE1_VALUE
          - VARIABLE2_NAME: VARIABLE2_VALUE

Note that, as a list, these are exported in order, which allows for later values to override (or modify) earlier values. Similarly, the `build-environment` from dependencies will be defined before that of the dependent's, allowing for it to be modified/overwritten by the dependent.

### CODE REVIEW

File: snapcraft/internal/deltas/_deltas.py [hunk 1]
```diff
@@ -159,7 +159,7 @@ def _update_progress_indicator(self, proc, progress_indicator):
                 count = 0
             progress_indicator.update(count)
             count += 1
-            time.sleep(.2)
+            time.sleep(0.2)
```
Comment (sergiusens) [inline snapcraft/internal/deltas/_deltas.py]: mind reverting this change?

File: snapcraft/internal/lifecycle/_packer.py [hunk 2]
```diff
@@ -102,7 +102,7 @@ def _run_mksquashfs(
                     count = 0
                 progress_indicator.update(count)
                 count += 1
-                time.sleep(.2)
+                time.sleep(0.2)
```
Comment (sergiusens) [inline snapcraft/internal/lifecycle/_packer.py]: ditto

File: tests/spread/general/build_environment/snaps/build-environment-test/snap/snapcraft.yaml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+summary: Quick snap to ensure build-environment works
+description: Quick snap to ensure build-environment works
+
+base: core18
+grade: devel
+confinement: strict
+
+apps:
+  build-environment-test:
+    command: echo "$FOO"
```
Comment (sergiusens) [inline tests/spread/general/build_environment/snaps/build-environment-test/snap/snapcraft.yaml]: wait, how is this possible?

Comment (kyrofa) [reply tests/spread/general/build_environment/snaps/build-environment-test/snap/snapcraft.yaml]: Oh it's not! At least, I hope it's not :grin: . This is here so we can [make _sure_ it's not](https://github.com/snapcore/snapcraft/pull/2322/files#diff-eb40a183cf74c2ef453f34aa1ebe4852R27).

File: snapcraft/internal/deltas/_deltas.py (hunk 1 above)
Comment (kyrofa) [reply snapcraft/internal/deltas/_deltas.py]: Sure thing, sorry-- new black issues.

File: snapcraft/internal/lifecycle/_packer.py (hunk 2 above)
Comment (kyrofa) [reply snapcraft/internal/lifecycle/_packer.py]: Also fixed.

**Pull request merged**